### DISCUSSION TITLE
Understanding rules evaluation

### DISCUSSION THREAD

**Discussion by AlekSi** [Mergifyio/mergify#5101] [General]
I'm having a hard time understanding what triggers rules evaluation and what triggers actions.

The only bit of information about that I was able to find is there: https://docs.mergify.com/workflow/writing-your-first-rule/#understanding-mergify-rule-conclusions

> Mergify continuously evaluates your rules whenever a pull request is updated or modified. When a rule's conditions are met, Mergify triggers the corresponding actions.

I think that "whenever a pull request is updated or modified" is a bit too vague. For example, that does not include edits of the PR title or description. It would be nice to have a list of things that trigger rules evaluation in the documentation.

Actions triggering is even more confusing. I [tried](https://github.com/FerretDB/merge-queues) a following configuration:

```yml
pull_request_rules:
  - name: "On conflict"
    conditions:
      - "conflict"
    actions:
      comment:
        message: "@{{author}} this pull request has merge conflicts."
```

I created a PR. Then, I pushed a commit to the base branch to create a merge conflict. The comment [was added](https://github.com/FerretDB/merge-queues/pull/1#issuecomment-1783770336) to the PR. So far, so good. But then I wasn't able to make this comment be added again by pushing a commit to either the base branch or PR branch. It seems like rules _are_ evaluated on pushes to the PR branch, but actions are executed only when rules previously evaluated to `false` and now evaluate to `true`. In other words, it is an edge-triggering system, not state-triggering. Is it true? Is it true for all rules, conditions, and actions? Again, it would be nice to have it documented.

**Comment by AlekSi** [comment]
(as a side note: typing `true` in the search box on https://docs.mergify.com makes the page disappear with `SyntaxError: JSON Parse error: Unrecognized token '<'` error in the console)

**Reply by jd** [reply]
Thanks, I've opened a bug on our side for fixing.

**Comment by jd** [comment]
> I think that "whenever a pull request is updated or modified" is a bit too vague. For example, that does not include edits of the PR title or description. It would be nice to have a list of things that trigger rules evaluation in the documentation.

It does. What makes you think it doesn't?

> I created a PR. Then, I pushed a commit to the base branch to create a merge conflict. The comment https://github.com/FerretDB/merge-queues/pull/1#issuecomment-1783770336 to the PR. So far, so good. But then I wasn't able to make this comment be added again by pushing a commit to either the base branch or PR branch. It seems like rules are evaluated on pushes to the PR branch, but actions are executed only when rules previously evaluated to false and now evaluate to true. In other words, it is an edge-triggering system, not state-triggering. Is it true? Is it true for all rules, conditions, and actions? Again, it would be nice to have it documented.

Yes, you're right; it's edge-triggered indeed (I never heard that expression but I'll be sure to include it in the docs 😄).
I'll make sure it's documented (I'm working on it).

**Reply by AlekSi** [reply]
> It does. What makes you think it doesn't?

I guess it was a problem in my configuration – it works now :) But also a lack of progress indicators. When I push new commits, I see the Summary PR status check as an orange dot before it becomes a green checkmark. When I change a title for the PR, for example, it stays the same green checkmark before suddenly triggering an action. I wonder if there is a race condition like in #5085. There are also no logs of evaluation in the events log dashboard.

**Reply by AlekSi** [reply]
There is also no attribute for the last editor of the title or description… I really wish you would implement a "wait" merge method from #5085.

**Reply by jd** [reply]
> I guess it was a problem in my configuration – it works now :) But also a lack of progress indicators. When I push new commits, I see the Summary PR status check as an orange dot before it becomes a green checkmark. When I change a title for the PR, for example, it stays the same green checkmark before suddenly triggering an action. I wonder if there is a race condition like in https://github.com/Mergifyio/mergify/discussions/5085. There are also no logs of evaluation in the events log dashboard.

This is because of the way GitHub checks work: they are not attached to a PR but to a sha1.

So when you push new commits, all the current checks disappear. To avoid that, Mergify pushes an "in-progress" indicator right away (orange circle), so you don't see Mergify totally disappearing, and a few seconds later, pushes back the Summary (green).

When you edit a PR, Mergify does not do anything right away but updates Summary a few seconds later. That's why you don't see any orange/in-progress status.

Evaluations are never logged, as they happen almost always when your repository has activity.

**Reply by AlekSi** [reply]
@jd Sorry, I think I still don't understand how that works. Consider the [following configuration](https://github.com/FerretDB/FerretDB/blob/7ce5098b030884bc69cf2580196ea8668452dc55/.github/mergify.yml#L24-L33):

```yaml
- name: "Request review"
  conditions:
    - "label = 'some/label'"
  actions:
    request_reviews:
      users: ["user1", "user2"]
      teams: ["@org/team1", "@org/team2"]
      users_from_teams: ["@org/team1", "@org/team2"]
    comment:
      message: "Some message."
```

Recently (today or a few days ago) Mergify started to re-request reviews from those users and teams after each review (see [there](https://github.com/FerretDB/FerretDB/pull/4512#event-13883905896)). The comment, however, was posted only once. Is it a new bug on the Mergify side? Or in my understanding?

**Reply by jd** [reply]
Remember Mergify is state-based, not event-based.
You wrote that if label `some/label` is present then a request for review for `user1`, etc needs to be present.

Mergify honors that. There is no conditions based on who might have already reviewed the PR in your configuration here.

**Reply by AlekSi** [reply]
So is it a recent bug or a recent fix? Maybe the definition of "request for review" changed recently? Because now it seems to mean "pending request for review", and before it seem to mean "request for review present in any state, including 'approved' and 'changes requested'". It seems like https://docs.mergify.com/workflow/request-reviews/#basic-assignment would fail now, too.

**Reply by jd** [reply]
Nothing changed as far as I can tell on Mergify's side (can't say for GitHub though), and the example still works as expected.
It just does not match what you expect, I guess.
Review states are not checked if you don't ask for them to be checked in `conditions`.

**Reply by AlekSi** [reply]
But… that configuration worked before for several months.

To put it another way: should [this example](https://docs.mergify.com/workflow/request-reviews/#basic-assignment) re-request review after each review? That's unlikely because that's pretty useless. But that example is not that different from my configuration.

**Reply by jd** [reply]
Yes, it does re-request review. We actually use it internally for months.

You're claiming something changed, but do you have an actual example of the previous behaviour that you saw?
I went ahead and looked at PR such as https://github.com/FerretDB/FerretDB/pull/3417 and it shows that Mergify request the review from "a team" each time a review is done on behalf of it.

**Reply by AlekSi** [reply]
/o\ Sorry, that was just me being blind. Thank you!
